Surface-tracker library functionality

romanroibu opened the pull request:
##### TODO

- [x] Define a surface within a frame
- [x] Locate a surface within a frame
- [x] Add/remove markers to/from a surface definition
- [x] Move a corner of a surface definition
- [x] Extract a crop of the surface location from an image
- [x] Calculate the surface heatmap based on a set of image points
- [x] Invalidate old surface locations after the surface definition was updated
- [ ] Serialize/deserialize surface definitions
- [ ] Serialize/deserialize surface locations

romanroibu committed:
Add basic python project structure

romanroibu committed:
Add primary public API definitions and basic tests

romanroibu committed:
Remove update method from GazeOnSurfaceDistribution to force batch calculation

romanroibu committed:
Move world_size property from Surface to GazeOnSurfaceDistribution

romanroibu committed:
Rename GazeOnSurfaceDistribution to GazeOnSurface

romanroibu committed:
Add mapping to/from image pixel space / surface normalized space

romanroibu committed:
Add (un)distort_points_to_image_plane methods to CameraModel

romanroibu committed:
Implement SurfaceLocation interface for version 2

Version 0 and 1 of the interface are/should be implemented in pupil core

romanroibu committed:
Move unit tests under tests/unit

romanroibu committed:
Create the structure for integration tests

romanroibu committed:
Add Visual Studio Code settings enabling Python Testing extension

romanroibu committed:
Add type annotation to utils.left_rotation and utils.right_rotation

romanroibu committed:
Fix numpy import in camera.py

romanroibu committed:
Explicitly specify the canonical order of the CornerIds

romanroibu committed:
Fix issue with location mapping from image space to surface space

romanroibu committed:
Refine the interface of SurfaceLocation

romanroibu committed:
Refine the Marker interface

romanroibu committed:
Refine Surface interface

romanroibu committed:
Add bounding_quadrangle helper function

romanroibu committed:
Add concrete marker implementations for markers in surface space and image space

romanroibu committed:
Add concrete implementation of the Surface interface

romanroibu committed:
Add concrete implementation of the SurfaceLocation interface

romanroibu committed:
Update SurfaceTracker to use internal API to provide public functionality

romanroibu committed:
Add SurfaceVisualAnchors class that provides convenience interface for visualizing surface locations

romanroibu committed:
Expose all the public interfaces from the surface_tracker package

romanroibu committed:
Add a note to add the OpenCV dependency

romanroibu committed:
Add an example Jupyter notebook showcasing the basic use of the library

romanroibu committed:
Run the Jupyter notebook showcasing the basic use of the library

romanroibu committed:
Apply black formatting

romanroibu committed:
Refine and extend the SurfaceTracker interface

romanroibu committed:
Move bounding_quadrangle and GetAnglesPolyline from utils into surface

romanroibu committed:
Update Surface interface to mark the methods as internal

romanroibu committed:
Update Marker interface to pass a corner order to the method returning the vertices

romanroibu committed:
Update SurfaceLocation to pass the corner order explicitly to Marker methods

romanroibu committed:
Add methods to SurfaceLocation to map a marker to/from image space from/to surface space

romanroibu committed:
Implement adding/removing a marker from the surface definition

romanroibu committed:
Implement moving a corner in the surface definition

romanroibu committed:
Implement getting corner position in SurfaceTracker

romanroibu committed:
Implement moving corner position in SurfaceTracker

romanroibu committed:
Implement adding/removing marker in SurfaceTracker

romanroibu committed:
Update visualization notebook with refined SurfaceTracker API and methods to add/remove markers and move corners

romanroibu committed:
Add requirement to CameraModel to implement resolution property

romanroibu committed:
Add SurfaceImageCrop class and related methods

romanroibu committed:
Add SurfaceHeatmap functionality replacing GazeOnSurface

romanroibu committed:
Add example code for getting the image crop and the points heatmap

marc-tonsen reviewed (approved):
Looks excellent! As discussed, we might want to make changes to how the heatmap is computed, but for the moment this is great!

romanroibu committed:
Fix issue wil calculating undistorted new corner position in surface space

romanroibu committed:
Extract method for transforming a list of points in surface space into points in image space

romanroibu committed:
Fix Surface.from_dict method

romanroibu committed:
Add method to get multiple corner positions to SurfaceTracker

romanroibu committed:
Add method to move multiple corners at once to SurfaceTracker

romanroibu committed:
Add is_stale property to SurfaceLocation to check if the computed location is stale relative to the surface definition

The location is considered stale if the surface definition changed after the location was computed.

romanroibu committed:
Add SurfaceTrackerWeakLocationStore and SurfaceTrackerArgumentValidator

romanroibu committed:
Replace add_marker_to_surface with add_markers_to_surface method in SurfaceTracker

romanroibu committed:
Replace remove_marker_from_surface with remove_markers_from_surface method in SurfaceTracker

romanroibu committed:
Add argument validation to all methods in SurfaceTracker

romanroibu committed:
Add location tracking, and marking as stale for methods that mutate the surface definition

romanroibu committed:
Update example notebook to latest API changes

romanroibu commented:
@s-westphal Please check the latest commits; they include the ability to add/remove multiple markers in one call, as well as moving multiple corners at once. There is an example notebook that shows how to use all the APIs. I'm happy to answer any further questions you might have.

romanroibu commented:
@papr Could you please take a brief look at this PR? If nothing jumps out as problematic, I think the branch can be safely merged.

s-westphal commented:
Amazing, thank you @romanroibu !!

papr commented on src/surface_tracker/marker.py:
```diff
@@ -0,0 +1,192 @@
+        try:
+            actual_space = value["space"]
+            expected_space = _MarkerInSurfaceSpace._COORDINATE_SPACE
+            if actual_space != expected_space:
```
Nice!

papr commented on src/surface_tracker/visual_anchors.py:
```diff
@@ -0,0 +1,87 @@
+        edit_surface_anchor = title_anchor[0], title_anchor[1] + 25
+        edit_markers_anchor = title_anchor[0], title_anchor[1] + 50
+
+        return SurfaceVisualAnchors(
```
Why not make it a classmethod and do this?
```suggestion
        return cls(
```

papr commented on src/surface_tracker/tracker.py:
```diff
@@ -0,0 +1,330 @@
+
+        # TODO: Check for markers uniqueness
+
+        # TODO: Check markers are not already part of the surface definition
```
What about these todos?

romanroibu replied:
Done

papr commented on src/surface_tracker/location.py:
```diff
@@ -0,0 +1,474 @@
+            registered_vertices_undistorted, visible_vertices_undistorted
+        )
+
+        return _SurfaceLocation_v2(
```
Similar to the other comment: Why not make this a classmethod and return
```suggestion
        return cls(
```
?

romanroibu replied:
In the case of `SurfaceLocation`, the `SurfaceLocation` class is a facade for concrete implementations.

`SurfaceLocation._create_location_from_markers` must return a concrete instance that implements the `SurfaceLocation` interface. The exact implementation that is returned here might change in the future (e.g. `_SurfaceLocation_v3`, etc.) and the only guarantee is that the returned instance implements the common interface.

Also, since the facade API implies calling the method on `SurfaceLocation`, making it a `classmethod` won't work, since `cls` in this case will be `SurfaceLocation`, which is an abstract class. The user would be required to explicitly call `_SurfaceLocation_v2._create_location_from_markers` which defeats the purpose of the facade in the first place.

Please let me know if this makes sense or not.

papr replied:
Thank you for the detailed explanation. In the context of the abstract interface, this makes indeed a lot of sense!

papr reviewed (commented):
I added a few comments but nothing blocking.

romanroibu committed:
Check markers to be added/removed to/from surface definition in SurfaceTracker
